### PR TITLE
feat: Update specification text for address book service

### PR DESCRIPTION
**Pull request by jsync-swirlds**
* Updated specification/API comments

### Rendered documentation
A rendered version of the documentation in these files is _temporarily_ available on [another branch](https://github.com/hashgraph/hedera-protobufs/tree/temp-DAB-Update-Spec-Backup/documents/api/services) to assist with visualizing what these changes accomplish.

### CODE REVIEW

File: services/address_book_service.proto
Comment (iwsimon) [inline services/address_book_service.proto:31]: administrators to add, update and remove consensus nodes. The addition, updating or removal

File: services/address_book_service.proto
Comment (iwsimon) [inline services/address_book_service.proto:53]: There is no "admin key" in these transactions.

Comment (jsync-swirlds) [reply services/address_book_service.proto:53]: The idea is to specify which key must sign.  Account only has one key attribute, however, so I'll reword and clarify.

File: services/address_book_service.proto
Comment (iwsimon) [inline services/address_book_service.proto:57]: When executing the next `freeze` transaction with `freeze_type` set to`PREPARE_UPGRADE`, config.txt, .pem file per each node, and a marker file will be generated.

File: services/address_book_service.proto
Comment (iwsimon) [inline services/address_book_service.proto:80]: Same as above.

File: services/address_book_service.proto
Comment (iwsimon) [inline services/address_book_service.proto:118]: Even this transaction is complete, the new node is still not joining the network until next freeze upgrade for phase 2 and not joining the network until consensus roster sent in the next few round for phase 3 (more details still pending to be narrowed down.)

File: services/node_create.proto
Comment (iwsimon) [inline services/node_create.proto:39]: We should remove "pending" state here.

File: services/address_book_service.proto
Comment (jsync-swirlds) [reply services/address_book_service.proto:34]: HIP text _appears to_ claim a node operator can update their own node *without* council approval. Is council approval required for everything (in which case node operators are never actually required), or are node operators permitted to update their own nodes independently?

File: services/address_book_service.proto
Comment (jsync-swirlds) [reply services/address_book_service.proto:57]: I prefer not to include internal functional details (when _completely_ invisible to external operations) in specification.
Stating what observable transaction causes a state change is helpful to an implementor or client; but including files written to local disk seems an unnecessary detail (another implementation may not do the same thing, it's not observable, and clients shouldn't care).

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto:39]: What state is the internal (in the merkle tree) entry created, if not "pending"?  I just copied that from the previous text and reworded.

File: services/address_book_service.proto
Comment (jsync-swirlds) [reply services/address_book_service.proto:118]: I'll clarify that the node is added to _state_ (which it must be, or the consensus roster will never be updated in a later process).

File: services/node_create.proto
Comment (iwsimon) [inline services/node_create.proto:82]: A list of grpc service endpoints. "grpc" needs to be specified in the previous notion discuss.

File: services/node_create.proto
Comment (iwsimon) [inline services/node_create.proto:85]: ip address instead of address

File: services/node_create.proto
Comment (iwsimon) [inline services/node_create.proto:86]: DNS name is optional, not "MUST NOT".

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto:82]: These are service endpoints.  gRPC should be specified in the requirements lines below if that is strictly required (not sure it truly is, but no harm adding it for now).

I'll add the gRPC requirement.

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto:86]: And yet, in another location we state the system will fail if DNS is specified.
We have to decide if this is allowed or not and make it clear.  I assumed the more restrictive of the two, as "this is optional, but your transaction will fail" is not what I call optional.

File: services/node_create.proto
Comment (iwsimon) [inline services/node_create.proto:97]: Keep this "Precisely, this field is the DER encoding of gossip X509 certificate."

Comment (jsync-swirlds) [reply services/node_create.proto:97]: That sentence does not make sense.
Also, the latest update in channel is that we must use PEM encoding, not DER.
Note, that this won't merge before the next discussion where additional updates for the HIP are likely.

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto:85]: I believe `address` is sufficient here.
Adding `IP` is unnecessary as the sub-type of address is defined by the `ServiceEndpoint` message.

File: services/node_create.proto
Comment (iwsimon) [reply services/node_create.proto:86]: For gossip_endpoint, fully qualified domain name is not allowed; for service_endpoint, fully qualified domain name is allowed.

Comment (jsync-swirlds) [reply services/node_create.proto:86]: Definitely some vagueness in HIP text for that.
I'll adjust the `service_endpoint` text to state the endpoints `MAY` specify a DNS name.

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto]: I've updated this text with a possible reword; it's a bit of a guess, so please review carefully.

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto]: Wording updated to flow better and be a bit more clear on formats and encoding.

File: services/address_book_service.proto
Comment (iwsimon) [inline services/address_book_service.proto:43]: All updates need approval.

File: services/node_create.proto
Comment (iwsimon) [reply services/node_create.proto:99]: Address could have different meanings. We have Address and AddressBook concepts in the code. We can use IP instead of address here.

File: services/node_create.proto
Comment (iwsimon) [reply services/node_create.proto]: DER is the encoding format to store the certificate. PEM is the file extension.

File: services/state/addressbook/node.proto
Comment (jsync-swirlds) [reply services/state/addressbook/node.proto:165]: The entire structure makes no sense if deletion has no status.  What you're describing is a non-functional half-state, which should not be specified.
What is the intention to manage deleted nodes; that is what we need to specify and we'll adjust that _if_ future phases choose a different approach.

File: services/address_book_service.proto
Comment (jsync-swirlds) [reply services/address_book_service.proto:43]: Repeating my earlier response:
HIP text appears to claim a node operator can update their own node without council approval. Is council approval required for everything (in which case node operators are never actually required), or are node operators permitted to update their own nodes independently?

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto]: PEM is absolutely a format, and it's not identical to DER (although the two _can be_ combined).
DER is a binary format (ASN.1 "Distinguished Encoding Rules") for encoding data, often used for PKI certificates (of various forms).
PEM is an ASCII format (S/MIME "Privacy Enhanced Mail") that Base64 encodes binary data as ASCII text with specific delimiters.
When the two are combined (which is unambiguous here) as a "PEM encoded certificate" you get a PKI certificate encoded to binary and then Base64 encoded to ASCII text and stored in a file with specified delimiters.

Specifying DER here with a "PEM" file extension would leave us with an unusable binary file that cannot be properly processed as a PEM format file and might result in data corruption if the file is assumed to be actual PEM format and treated as 7-bit-safe ASCII.

Comment (iwsimon) [reply services/node_create.proto]: "Precisely, this field is the DER encoding of gossip X509 certificate." This was Nathan's words to keep during our team meeting together. We parse .PEM file and get X509Certificate and store it's getEncoded() which is DER.

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto:99]: Context is already clear here, a ServiceEndpoint (the type of _this_ field) only holds IP addresses, and the wording "address and port" is the common way of differentiating from FQDN.

File: services/state/addressbook/node.proto
Comment (jsync-swirlds) [reply services/state/addressbook/node.proto:165]: Based on discussion, DELETED state is removed (the node will be removed entirely, although we'll still need to ensure node ID is never reused).
We still need the two `PENDING` status, without that we have no clear marker for what to change in the active configuration at the next upgrade.

This entity is internal network state (and part of the implementation details), so adding a status to enable managing the items modified but not yet active in network configuration is reasonable, in my opinion.

File: services/address_book_service.proto
Comment (iwsimon) [reply services/address_book_service.proto:34]: From today's meeting, update don't need council approval.

File: services/node_create.proto
Comment (iwsimon) [inline services/node_create.proto:86]: Don't want to loss these details:
The first in the list is used as the Internal IP address in config.txt,
 the second in the list is used as the External IP address in config.txt,
  the rest of IP addresses are ignored for DAB phase 2.

Comment (jsync-swirlds) [reply services/node_create.proto:86]: Those are specific to our implementation, and would belong in a design document, but not here in the protobuf.
I can note the internal/external part (not config.txt, however) if we believe it will remain a _requirement_ for all implementations.

File: services/query.proto
Comment (iwsimon) [inline services/query.proto:207]: Don't need the whole package name here.

File: services/state/addressbook/node.proto
Comment (iwsimon) [reply services/state/addressbook/node.proto:167]: We should remove this enum NodeStatus. It's not used in phase 2. It was designed for phase 3 to deal with Roster. These comments do not make sense. We can add them later for phase 3 if needed.

File: services/node_create.proto
Comment (iwsimon) [reply services/node_create.proto:86]: I will copy them to java file comments.

File: services/query.proto
Comment (jsync-swirlds) [reply services/query.proto:207]: It is required now; as the package directive in the files is updated.

File: services/state/addressbook/node.proto
Comment (jsync-swirlds) [reply services/state/addressbook/node.proto:167]: Not sure how the comments don't make sense; they describe the behavior specified elsewhere, just specific to the status.
We aren't going to be able to implement, even phase 2, without some sort of status for pending items.
It might have been missed, but something has to track that a node is added in state but not written to configuration yet, and that something has to be _in_ state.

Comment (iwsimon) [reply services/state/addressbook/node.proto:182]: In phase 2, we use a store to track those hapi transactions and they are not active until next freeze update. Here is the design [diagram](https://lucid.app/lucidchart/01f7f0ae-a6e3-4e44-9768-9220936ecec6/edit?invitationId=inv_bb15f582-8437-496f-875c-3139d5bdadfe&page=m-5o7ONTd-nK#).

Comment (jsync-swirlds) [reply services/state/addressbook/node.proto:182]: From that diagram the "store" (which is a proxy for the merkle tree) is storing `Node` objects (exactly the message that we're defining here).
Nothing in that diagram there explains how the freeze upgrade process will know _which_ `Node` objects need to be _updated_, _added_, or _deleted_.
Pending status values (and one non-pending value) seem, to me, to be the most sensible mechanism for such indicator.

File: services/state/addressbook/node.proto
Comment (netopyr) [inline services/state/addressbook/node.proto:143]: We should not define changes for an incomplete design.

File: services/state/addressbook/node.proto
Comment (netopyr) [inline services/state/addressbook/node.proto:153]: We should not define `NodeStatus` at this point. The design for the next phase is incomplete, and one can easily imagine a solution that does not need a `NodeStatus`.

**Review by netopyr — CHANGES_REQUESTED** [review]
Please remove the `status` property and `NodeStatus`. The design for this functionality is incomplete and the usage of `status` and `NodeStatus` is not a given.

File: services/state/addressbook/node.proto
Comment (jsync-swirlds) [reply services/state/addressbook/node.proto:153]: What you are requesting is a non-functional design.  Without *something* we have no pending state and much of the rest of the design is impossible to implement.  That suggests that *this phase* design is incomplete, in which case we need more work and should consider the need for some form of status(or alternative) as part of the current phase.

File: services/transaction_body.proto
Comment (jsync-swirlds) [reply services/transaction_body.proto:333]: This reorder moves the transactions for each service to be grouped together (so token service transactions are together, crypto service transaction are together, etc...); otherwise services end up scattered about, with some transactions for the same service quite far apart.

It could be returned to number order, but that, in my opinion, makes the specification less readable for an outside contributor.

Perhaps something to discuss more widely, as number order is something more helpful when _adding_ items, but service-grouped is more helpful when _finding_ items.

File: services/address_book_service.proto
Comment (jsync-swirlds) [reply services/address_book_service.proto:77]: Validate the signature on the transaction?
Not sure what isn't possible here...

File: services/basic_types.proto
Comment (jsync-swirlds) [reply services/basic_types.proto:1389]: I believe so, as this is a field name in the same message, so it's a code reference.
The guideline I'm following is to use inline code styling (\`) for references to field or message names.

File: services/basic_types.proto
Comment (jsync-swirlds) [reply services/basic_types.proto:1378]: To add a line break after the sentence (otherwise markdown will merge the lines).
It could be a `<p>` tag (technically `<p/>` is more correct, but that confuses javadoc), and in future (much more thorough edits) that is under consideration for the standard separation between description and constraint.

File: services/state/addressbook/node.proto
Comment (jsync-swirlds) [reply services/state/addressbook/node.proto:153]: Note, if we have a solid design that will definitely work _as specified_ for phase 2; I have no issue to remove this pre-existing definition entirely. If we intend to keep this in the protobuf definitions, however, it needs to be well specified.

Side note. If we have a design that works but is _different_ from specification, then, in my opinion, we need to **publicly document** that design and then we can update specification accordingly.

File: services/state/addressbook/node.proto
Comment (jsync-swirlds) [reply services/state/addressbook/node.proto]: Updated based on discussion.

File: services/state/addressbook/node.proto
Comment (jsync-swirlds) [reply services/state/addressbook/node.proto]: Status field and enum are removed.

File: services/node_create.proto
Comment (edward-swirldslabs) [inline services/node_create.proto:87]: I don't know if this impacts what is written here, in the code the effective difference between "internal" and "external" is the order in which the connection attempts are made.  If the "internal" address resolves, we attempt to connect to that address first.  Once we start having consensus rosters, the generalized treatment will be to attempt connection in order of the list and not make a categorical discrepancy between "internal" and "external".

File: services/node_create.proto
Comment (edward-swirldslabs) [inline services/node_create.proto:119]: Right now I believe all signing certs are self-signed.  I don't know if the self-signed property makes a difference on its ability to be used in the trust store for the agreement keys in mutual TLS.  We will be automatically rendering the mTLS agreement keys on boot with the signing cert as a parent certificate.  If the self-signed property is necessary for it to function as the parent cert of the mTLS key, then this should be documented here.

**Review by edward-swirldslabs — APPROVED** [review]

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto:119]: Using a self-signed signing certificate as parent for the mTLS agreement key setup is neither required nor disbarred, it is just a matter of how you establish trust of the certificate and what authenticity guarantees the network is willing to accept.

That's just standard mTLS considerations.

File: services/node_create.proto
Comment (jsync-swirlds) [reply services/node_create.proto:87]: That will be, as I understand it, something visible in config.txt; and does not affect what is stored here in this phase.
Each `Node` from the in-state store will become an entry in `config.txt` as part of the freeze/upgrade process, and certificates (with their public keys) will be written to PEM files on disk.
As to later, the network address book (as distinct from the internal platform concept of the "roster") will need to track more than just consensus nodes, and will be used by more than just consensus nodes. Changes and/or adjustments needed for those later requirements are deferred to later consideration (c.f. Michael Heinrichs' comments above).

File: services/address_book_service.proto
Comment (netopyr) [reply services/address_book_service.proto:77]: I am not sure if validating a key _if present_ is possible. We can define the account key as optional, but this will only tell us whether there is a valid signature or not. We cannot distinguish between an invalid signature and a missing signature. I believe this distinction would be required here.

File: services/basic_types.proto
Comment (netopyr) [reply services/basic_types.proto:1389]: I find the mixture of html and markdown for formatting confusing. But if this is the standard we want to follow...

**Review by netopyr — COMMENTED** [review]
Please note that we have to pull merged changes for the dynamic address book out of `main` and keep them in a branch. They need to be brought back once we are ready to release the functionality.

File: services/address_book_service.proto
Comment (jsync-swirlds) [reply services/address_book_service.proto:77]: If the signature is present, we can certainly validate that, and if not present not validate the signature.  I think I see your point; the wording is a bit confusing, then again the process is a bit confusing too (because it might be initiated in two different ways).
I'll rework the wording a bit to make the necessary actions a bit clearer.

File: services/basic_types.proto
Comment (jsync-swirlds) [reply services/basic_types.proto:1389]: It's all markdown; it's just that base markdown has two large gaps normally filled with html tags.
1. Standard markdown has inconsistent support for a line break.  It adds a paragraph break for each blank line, but single line break is recommended to use `<br/>` for best compatibility.
1. Markdown requires the use of HTML tags for anything multi-line in tables because tables are an afterthought in markdown so a table row cannot contain a newline.
What the processor does is remove all line breaks in these comment blocks before building the table row, so we have to use tags a bit more in field comments than in message comments.
The main alternative would be to use html exclusively, but that's much harder to read, in my opinion, than markdown.

**Comment by jsync-swirlds** [comment]
> Please note that we have to pull merged changes for the dynamic address book out of `main` and keep them in a branch. We need to bring them back once we are ready to release the functionality. We should retarget this PR against this branch once it was created.

Why would we remove these from `main`?  This is an API definition, the definition is complete for phase 2, and there is no risk to publishing the API prior to completing the code design.

File: services/address_book_service.proto
Comment (iwsimon) [inline services/address_book_service.proto:55]: Maybe remove '"pending state' here, same below 'transition the new pending node to an active status'? We can say "the new node is not active until next freeze updrade' or something like that.

Comment (jsync-swirlds) [reply services/address_book_service.proto]: Updated here, and also for the similar items in update and delete.

**Review by iwsimon — DISMISSED** [review]
LGTM. Thanks @jsync-swirlds

**Comment by jsync-swirlds** [comment]
Note, merge for this PR is delayed to NET 2024-05-31 to accommodate a requirement for release `0.51`.

**Review by Mark-Swirlds — COMMENTED** [review]
(small comment) Can we change the name of node_create to node_add?  I understand the CRUD structure, but node_add is more descriptive of the task.

**Comment by jsync-swirlds** [comment]
> (small comment) Can we change the name of node_create to node_add? I understand the CRUD structure, but node_add is more descriptive of the task.

1. @iwsimon would be the best individual to do that in the PR to add all these files back into the protobuf repo (everything related to DAB was removed from both protobuf and services codebase on Wednesday).
1. The HIP would need to be rewritten to change all the `create` to `add` wording.
1. That would be inconsistent with every other service use of `create` for adding or creating an entity
1. It is creating an entity in state, so `create` seems entirely reasonable to my mind.

**Comment by jsync-swirlds** [comment]
Updated following the removal and re-add of all address book changes.
That bounce required squashing these commits to make rebase sane.

**Comment by jsync-swirlds** [comment]
> Please remove the `status` property and `NodeStatus`. The design for this functionality is incomplete and the usage of `status` and `NodeStatus` is not a given.

@netopyr This is done, along with matching the most recent discussions and updates.  Please re-review.

**Review by netopyr — COMMENTED** [review]
LGTM - thanks @jsync-swirlds

**Review by netopyr — APPROVED** [review]

**Review by JeffreyDallas — APPROVED** [review]

**Review by povolev15 — APPROVED** [review]
LGTM

**Pull request merged**
